apps/www/_blog/2024-08-13-supabase-realtime-broadcast-and-presence-authorization.mdx

```diff
@@ -0,0 +1,208 @@
+
+We decided on an approach that keeps your database and RLS policies at the heart of this new authorization strategy.
+
+### Database as a source of security
+
+To achieve Realtime authorization, we looked into our current solutions, namely how [Storage](https://supabase.com/docs/guides/storage/security/access-control) handles Access Control. Due to the nature of Realtime, our primitives are different as we have no assets stored in the database. So how did we achieve it?
+
+On Channel subscription you are able to inform Realtime to use a private Channel and we will do the required checks.
+
+The checks are done by running SELECT and INSERT queries on the new `realtime.messages` table which are then rolled backed so nothing is persisted. Then, based on the query result, we can determine the policies the user has for a given extension.
+
+As a result, in the server, we create a map of policies per connected socket so we can keep them in memory associated with the user's connection.
+
+```elixir
+%Policies{
+  broadcast: %BroadcastPolicies{read: false, write: false},
+  presence: %PresencePolicies{read: false, write: false}
+}
+```
+
+### One user, one context, one connection
+
+Now that we have set up everything on the database side, let's understand how it works and how we can verify authorization via RLS policies.
+
+Upon channel subscription, based on the configuration message sent when joining the websocket, we determine that you want to use a private channel, then we parse the information from your signed JSON Web Token (JWT), load them into a Postgres transaction using `set_config`, and verify them against the new `realtime.messages`table within the same transaction and we store the output as a group of policies within the context of the user's channel process.
+
+<Img
+  src={{
+    dark: '/images/blog/2024-08-13-supabase-realtime-broadcast-and-presence-authorization/one-user-connection--dark.png',
+    light:
+      '/images/blog/2024-08-13-supabase-realtime-broadcast-and-presence-authorization/one-user-connection--light.png',
+  }}
+  alt="Flow of checks done against the databse to determine the policies for a user connection."
+/>
+## How does this scale?
+
+Since we're doing this as soon as the user subscribes to Channel we should see a small latency increase in subscriptions but after that we do not run any queries on the database and we use the stored context to control the actions a user can perform.
+
+Latency between geographically close users is very important for a product like Realtime. To deliver messages as fast as possible between users on our global network, we cache the policies.
+
+We can maintain high throughput and low latency on a Realtime Channel with Broadcast and Presence authorization because:
+
+- the policy is only generated when a user connects to a Channel
+- the policy cached in memory is close to your users
+- the policy is cached for the duration of the connection or until the JWT expires
```

Comment:
it's very confusing when policy is mentioned here b/c i'm assuming you're talking about the Realtime cached policy. there's also RLS policy and i don't think we've actually clearly defined the difference between RLS policies and cached policies.

for example, up above there's "we create a map of policies per connected socket" and included this as an example:

```elixir
%Policies{
  broadcast: %BroadcastPolicies{read: false, write: false},
  presence: %PresencePolicies{read: false, write: false}
}
```

however, after that it's very hard to tell the difference.

We should have first established that there are RLS policies and Realtime cached policies that map to RLS policies or something.

Reply:
I think this is fine, we're basically describing that we're being careful on how we create the cache near the user vs far away which highlights the fact that we are aware of geography and take that into consideration making realtime look better for larger scale imo